discovery_engine_core/web-api/openapi.yaml

```diff
@@ -0,0 +1,198 @@
+    get:
+      tags:
+        - personalization
+      summary: Get personalized documents for the user
+      description: |-
+        Returns a map from `document_id` to the score of that document for the
+        specified `user_id`. The score is a value between 0 and 1 where a higher value
+        means that the document matches the preferences of the user better.
+      parameters:
+        - name: user_id
+          in: path
+          description: Id of the user
+          required: true
+          schema:
+            $ref: '#/components/schemas/UserId'
+        - name: count
+          in: query
+          description: Maximum number of personalized documents to return
+          required: false
+          schema:
+            type: integer
+            format: int32
+            minimum: 1
+            maximum: 100
+            default: 10
+      responses:
+        '200':
+          description: successful operation
+          content:
+            application/json:
+              schema:
+                $ref: '#/components/schemas/PersonalizedDocumentsResponse'
+        '400':
+          description: invalid user id.
+        '404':
+          description: user not found.
+        '422':
+          description: impossible to create a personalized list for the user.
+          content:
+            application/json:
+              schema:
+                $ref: '#/components/schemas/PersonalizedDocumentsError'
+
+  /users/{user_id}/documents/interact:
+    patch:
+      tags:
+        - personalization
+      summary: Add interaction between a user and a document
+      description: |-
+        The interaction is used to provide personalized documents to the user.
+      parameters:
+        - name: user_id
+          in: path
+          description: Id of the user
+          required: true
+          schema:
+            $ref: '#/components/schemas/UserId'
+      requestBody:
+        content:
+          application/json:
+            schema:
+              $ref: '#/components/schemas/UserInteractionRequest'
+      responses:
+        '204':
+          description: successful operation
+        '400':
+          description: invalid request. User or document id is invalid
+          content:
+            application/json:
+              schema:
+                $ref: '#/components/schemas/UserInteractionError'
+
+  /documents:
+    post:
+      tags:
+        - ingestion
+      summary: Add documents to the system
+      description: |-
+        Add documents to the system. The system will create a representation of the document
+        that will be used to match it against the preferences of a user.
+      requestBody:
+        content:
+          application/json:
+            schema:
+              $ref: '#/components/schemas/IngestionRequest'
+      responses:
+        '204':
+          description: successful operation
+        '400':
+          description: invalid request
+        '422':
```

Comment:
Should this be a 500 instead? Most probably is not a problem with the data that the user sent but something broke in the smbert model.
@rustonaut @biernacki @Robert-Steiner

Reply:
I mean it could be a problem with there data, like empty or single letter strings for snippets or similar.

If it's a bug that we thought there is a problem with a document it's still 422, as the bug is in the classification of mal-formedness not the reporting/handling of malformed documents.

If our models messes up it would be a bug and as such a 500 error code, but in this case we don't need to document it as 500 is pretty much implicitly always possible. 

Also I think I have seen some  request libraries which handle 500 errors different then 400 errors making conditional control-flow based on 500 errors hard. As such we should avoid them as long as we don't have a panic or error we believe is "bug-like" wrt. the full system (e.g. the database being unavailable is bug-like wrt. the system even through it's not a bug per-see)

Reply:
> Most probably is not a problem with the data that the user sent but something broke in the smbert model.
If the problem is not the user data but something else in our service, we should return 500.

But as @rustonaut wrote, I'm not sure it's worth mentioning here

Reply:
An empty strings will return a 400 we are returning that only if the embedding or the push to elastic search fails. I'll change it to 500

Reply:
I don't think that an smbert error is really expected.

Reply:
What if the string is non empty but only contains character we remove when normalizing?

I think a good idea could be deserialization failing => 400 no documents are  processed; post-deserialization validation failing => 422 only the specific document fails.

As we probably still want to ingest other documents even if one or two of them are bad I would say empty strings should be post-deserialization validation and yield 422 :thinking:

Reply:
Also internal server errors (500) are basically bugs. Besides basic robustness you normally don't write code which expects a bug. If we expect it to fail with some input then it's not a bug anymore and in turn not 500. If we don't expect it to fail its not necessary document or document that we return a list of failed documents.

Reply:
if that is the case then 500 works for me

Reply:
There is no normalization and there are no known inputs that will make smbert fail. If the tokenization or the run of the model fail is a bug in our code or in the libraries we are running.